File: 算法/排序/并归排序.py
def merge_sort(unordered_ls: list) -> list:
    if len(unordered_ls) <= 1:
        return unordered_ls

    # 将列表分成两半
    mid = len(unordered_ls) // 2
    left_half = unordered_ls[:mid]
    right_half = unordered_ls[mid:]

    # 递归对左右两半进行归并排序
    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)

    # 合并两个已排序的子列表
    return merge(left_half, right_half)


def merge(left: list, right: list) -> list:
    result = []     # 存储排序好的结果
    i = j = 0

    # 获取两个数组中小的部分排在 result list 里面
    # 获取完其中一个数组就结束
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # 这里写的非常巧妙，总有一个排序完了，所以我们获取其余部分是不会新增东西的
    # 没有排完的正好停在没有排序的位置，我们剩下的部分补充到数组中去
    result.extend(left[i:])
    result.extend(right[j:])
    return result


if __name__ == '__main__':
    arr = [12, 11, 13, 5, 6, 7]
    sorted_arr = merge_sort(arr)
    print(sorted_arr)

File: 算法/排序/并归排序.cpp
#include <iostream>
#include <vector>

using namespace std;

void merge(vector<int> &v, int l, int mid, int r) {
    int n1 = mid - l + 1;
    int n2 = r - mid;

    vector<int> vl(n1);
    vector<int> vr(n2);

    for (int i = 0; i < n1; ++i)
        vl[i] = v[l + i];

    for (int i = 0; i < n2; ++i)
        vr[i] = v[mid + 1 + i];

    int i = 0, j = 0, k = l;    // left 是每一次合并数组的最左边，我们依次排过来

    // 挑选两个数组中小的部分排序到前面
    while (i < n1 and j < n2) {
        // 这里控升｜降序
        if (vl[i] < vr[j]) {
            v[k] = vl[i];
            i++;
        } else {
            v[k] = vr[j];
            j++;
        }
        k++;
    }

    // 总有一个数组没有排序我完，然后排序那个数组
    while (i < n1) {
        v[k] = vl[i];
        i++;
        k++;
    }

    while (j < n2) {
        v[k] = vr[j];
        j++;
        k++;
    }
}

void merge_sort(vector<int> &v, int l, int r) {
    if (l < r) {
        int mid = l + (r - l) / 2;
        merge_sort(v, l, mid);
        merge_sort(v, mid + 1, r);
        merge(v, l, mid, r);
    }
}

int main() {
    vector<int> v = {10, 20, 40, 80, 5, 14, 30, 25};
    int n = static_cast<int>(v.size()) - 1; // 修正这里的数组长度

    merge_sort(v, 0, n);

    for (const auto &num: v)
        cout << num << " ";
    cout << endl;

    return 0;
}
